internal/auth/auth.go

```diff
@@ -0,0 +1,305 @@
+		tokens := &tokenSet{}
+		return tokens, json.Unmarshal(body, tokens)
+	}
+
+	tokenErrorBody := requestTokenError{}
+	if err := json.Unmarshal(body, &tokenErrorBody); err != nil {
+		return nil, errors.Wrapf(
+			err,
+			"unable to unmarshal requestTokenError from body %s",
+			body,
+		)
+	}
+	return nil, errors.Errorf(
+		"refreshing access token returned an error (%s) with description: %s",
+		tokenErrorBody.Error,
+		tokenErrorBody.ErrorDescription,
+	)
+}
+
+func (a *Authenticator) tryRequestToken(
+	codeResponse *codeResponse,
+) (*http.Response, error) {
+	reqURL := fmt.Sprintf("https://%s/oauth/token", a.Domain)
+
+	grantType := "urn:ietf:params:oauth:grant-type:device_code"
+	payload := strings.NewReader(fmt.Sprintf(
+		"grant_type=%s&device_code=%s&client_id=%s",
+		url.QueryEscape(grantType),
+		codeResponse.DeviceCode,
+		a.ClientID,
+	))
+
+	req, err := http.NewRequest(http.MethodPost, reqURL, payload)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	req.Header.Add("content-type", "application/x-www-form-urlencoded")
+
+	return http.DefaultClient.Do(req)
+}
+
+// handleFailure handles the failure scenarios for requestTokens.
+func handleFailure(body []byte, code int) (int, error) {
+	tokenErrorBody := requestTokenError{}
+	if err := json.Unmarshal(body, &tokenErrorBody); err != nil {
+		return 0, errors.WithStack(err)
+	}
+
+	if code == http.StatusTooManyRequests {
+		if tokenErrorBody.Error != "slow_down" {
+			return 0, errors.Errorf(
+				"got status code: %d, response body: %s",
+				code,
+				body,
+			)
+		}
+
+		return additionalSleepOnSlowDown, nil
+
+	} else if code == http.StatusForbidden {
+
+		// this error is received when waiting for user to take action
+		// when they are logging in via browser. Continue polling.
+		if tokenErrorBody.Error != "authorization_pending" {
+			return 0, errors.Errorf(
+				"got status code: %d, response body: %s",
+				code,
+				body,
+			)
+		}
+
+		return 0, nil // No slowdown, just keep trying
+	}
+	// The user has not authorized the device quickly enough, so
+	// the `device_code` has expired. Notify the user that the
+	// flow has expired and prompt them to re-initiate the flow.
+	// The "expired_token" is returned exactly once. After that,
+	// the dreaded "invalid_grant" will be returned and device
+	// must stop polling.
+	if tokenErrorBody.Error == "expired_token" || tokenErrorBody.Error == "invalid_grant" {
+		return 0, usererr.New(
+			"The device code has expired. Please try `devbox auth login` again.")
+	}
+
+	// "access_denied" can be received for:
+	// 1. user refused to authorize the device.
+	// 2. Auth server denied the transaction.
+	// 3. A configured Auth0 "rule" denied access
+	if tokenErrorBody.Error == "access_denied" {
+		return 0, usererr.New("Access was denied")
+	}
+
+	// Unknown error
+	return 0, usererr.New("Unable to login")
+}
+
+func getAuthFilePath() string {
+	return xdg.StateSubpath(filepath.FromSlash("devbox/auth.json"))
+}
```

Comment:
is this mostly taken from the `jetpack` tool login code? what changes were made?

Reply:
this stuff is new (launchpad doesn't store auth in state XDG I don't think, it also has a custom struct instead of just storing the auth0 response). The device code polling is inspired by launchpad but heavily modified in order to be simpler and not require a lot of copy pasting or importing (i.e. protos from that repo)